File: python/111-120/Populating Next Right Pointers in Each Node II.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return
    
        while True:
            cur = root
            pre = None
            while cur:
                if cur.left:
                    if not pre:
                        pre = cur.left
                        root = pre
                    else:
                        pre.next = cur.left
                        pre = pre.next
            
                if cur.right:
                    if not pre:
                        pre = cur.right
                        root = pre
                    else:
                        pre.next = cur.right
                        pre = pre.next
        
                            cur = cur.next
                                
                                if not pre:
                                break

File: cpp/111-120/Populating Next Right Pointers in Each Node II.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        while (root) {
            TreeLinkNode *tln = root;
            TreeLinkNode *prev = new TreeLinkNode(0);
            TreeLinkNode *temp = prev;
            
            while (tln) {
                
                if (tln->left) {
                    prev->next = tln->left;
                    prev = prev->next;
                }
                
                if(tln->right) {
                    prev->next = tln->right;
                    prev = prev->next;
                }
                
                tln = tln->next;
            }
            
            root = temp -> next;
        }
    }
};
